python/integration_tests/test_upkeep_retry.py

```diff
@@ -0,0 +1,337 @@
+
+import yaml
+
+from collections import defaultdict
+from python.integration_tests.helpers import (
+    TASKBROKER_BIN,
+    TESTS_OUTPUT_ROOT,
+    send_messages_to_kafka,
+    create_topic,
+)
+
+from python.integration_tests.worker import ConfigurableTaskWorker, SimpleTaskWorker, TaskWorkerClient
+
+
+TEST_OUTPUT_PATH = TESTS_OUTPUT_ROOT / "test_upkeep_retry"
+
+
+class TasksRetriedCounter:
+    """
+    A thread safe class that tracks the total number of tasks that have
+    been retried.
+    """
+    def __init__(self):
+        self.total_retried = 0
+        self.tasks_retried = defaultdict(int)  # key: task_name, value: number of retries
+        self._lock = threading.Lock()
+
+    def increment(self, task_name: str):
+        with self._lock:
+            self.total_retried += 1
+            self.tasks_retried[task_name] += 1
+            return self.total_retried
+
+    def get_total_retried(self):
+        with self._lock:
+            return self.total_retried
+
+    def get_tasks_retried(self):
+        with self._lock:
+            return self.tasks_retried
+
+
+counter = TasksRetriedCounter()
+
+
+def manage_taskworker(
+    worker_id: int,
+    consumer_config: dict,
+    log_file_path: str,
+    tasks_written_event: threading.Event,
+    shutdown_event: threading.Event,
+    num_messages: int,
+    retries_per_task: int,
+) -> None:
+    print(f"[taskworker_{worker_id}] Starting taskworker_{worker_id}")
+    worker = ConfigurableTaskWorker(TaskWorkerClient(f"127.0.0.1:{consumer_config['grpc_port']}"), retry_rate=1)
+    retried_tasks = 0
+    next_task = None
+    task = None
+
+    # Wait for consumer to initialize sqlite and write tasks to it
+    print(
+        f"[taskworker_{worker_id}]: Waiting for consumer to initialize sqlite and write tasks to it..."
+    )
+    while not tasks_written_event.is_set() and not shutdown_event.is_set():
+        time.sleep(1)
+
+    if not shutdown_event.is_set():
+        print(f"[taskworker_{worker_id}]: Processing tasks...")
+    try:
+        while not shutdown_event.is_set():
+            if next_task:
+                task = next_task
+                next_task = None
+            else:
+                task = worker.fetch_task()
+
+            # If there are no more pending task to be fetched, check if all tasks have been retried.
+            # If so, shutdown the taskworker. If not, wait for upkeep to re-produce the task to kafka.
+            if not task:
+                task_retried_count = counter.get_total_retried()
+                if task_retried_count >= num_messages * retries_per_task:
+                    print(
+                        f"[taskworker_{worker_id}]: Total tasks retried reached: {task_retried_count}/{num_messages * retries_per_task}. Shutting down taskworker_{worker_id}"
+                    )
+                    shutdown_event.set()
+                    break
+                else:
+                    time.sleep(1)
+                    continue
```

Comment:
In the event that not all tasks are processed should we have a max time we'll wait for completion?

Reply:
We should! Added a timeout for taskworkers.